File: src/main/java/jk/kamoru/flayground/flay/service/BatchService.java
package jk.kamoru.flayground.flay.service;

import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.lang.ProcessBuilder.Redirect;
import java.nio.charset.Charset;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.StandardOpenOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Date;
import java.util.List;
import java.util.Map.Entry;
import java.util.stream.Collectors;

import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.BooleanUtils;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import jk.kamoru.flayground.FlayConfig;
import jk.kamoru.flayground.flay.domain.Flay;
import jk.kamoru.flayground.flay.source.FlaySource;
import jk.kamoru.flayground.info.domain.History;
import jk.kamoru.flayground.info.domain.History.Action;
import jk.kamoru.flayground.info.service.HistoryService;
import jk.kamoru.flayground.notice.service.NotificationService;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
public class BatchService {

	public static enum Option {
		/** moveWatchedVideo */ W, /** deleteLowerRankVideo */ R, /** deleteLowerScoreVideo */ S;
	}
	
	public static enum Operation {
		/** moveWatchedVideo */ W, /** deleteLowerRankVideo */ R, /** deleteLowerScoreVideo */ S,
		/** InstanceVideoSource */ I, /** ArchiveVideoSource */ A, /** Backup */ B
	}
	
	public static final String BACKUP_INSTANCE_JAR_FILENAME = "flayground-instance.jar";
	public static final String BACKUP_ARCHIVE_JAR_FILENAME  = "flayground-archive.jar";
	public static final String BACKUP_INSTANCE_CSV_FILENAME = "flay-instance.csv";
	public static final String BACKUP_ARCHIVE_CSV_FILENAME  = "flay-archive.csv";

	@Autowired FlaySource instanceFlaySource;
	@Autowired FlaySource  archiveFlaySource;
	@Autowired HistoryService historyService;
	@Autowired NotificationService notificationService;

	@Value("${batch.watch.move}")   Boolean moveWatched;
	@Value("${batch.rank.delete}")  Boolean deleteLowerRank;
	@Value("${batch.score.delete}") Boolean deleteLowerScore;
	
	@Value("${path.video.archive}") String   archivePath;
    @Value("${path.video.storage}") String   storagePath;
    @Value("${path.video.stage}")   String[]   stagePaths;
    @Value("${path.video.cover}")   String[]   coverPaths;
    @Value("${path.video.queue}")   String     queuePath;
	@Value("${path.info}")          String      infoPath;
    @Value("${path.backup}")        String    backupPath;
    @Value("${path.video.stage},${path.video.cover}") String[] emptyManagedInstancePath;
    
    @Value("${size.storage}") int storageGbSize;
	
	public void reload() {
		log.info("[reload]");
		instanceFlaySource.load();
		notificationService.announce("Reload", "Instance Source");
	}

	public Boolean getOption(Option type) {
		switch (type) {
		case W:
			return moveWatched;
		case R:
			return deleteLowerRank;
		case S:
			return deleteLowerScore;
		default:
			throw new IllegalArgumentException("unknown batch option");
		}
	}

	public Boolean toggleOption(Option type) {
		switch (type) {
		case W:
			return moveWatched = BooleanUtils.negate(moveWatched);
		case R:
			return deleteLowerRank = BooleanUtils.negate(deleteLowerRank);
		case S:
			return deleteLowerScore = BooleanUtils.negate(deleteLowerScore);
		default:
			throw new IllegalArgumentException("unknown batch option");
		}
	}

	public void startBatch(Operation oper) {
		switch (oper) {
		case W:
			moveWatched();
			break;
		case R:
			deleteLowerRank();
			break;
		case S:
			deleteLowerScore();
			break;
		case I:
			instanceBatch();
			break;
		case A:
			archiveBatch();
			break;
		case B: 
			backup();
			break;
		default:
			throw new IllegalArgumentException("unknown batch operation");
		}
		
	}

	private void instanceBatch() {
		if (moveWatched)
			moveWatched();
		if (deleteLowerRank)
			deleteLowerRank();
		if (deleteLowerScore)
			deleteLowerScore();
		
		assembleFlay();
		deleteEmptyFolder(emptyManagedInstancePath);
		instanceFlaySource.load();
		notificationService.announce("Batch", "Instance Source");
	}

	private void archiveBatch() {
		relocateArchiveFile();
		deleteEmptyFolder(archivePath);
		archiveFlaySource.load();
		notificationService.announce("Batch", "Archive Source");
	}

	private void relocateArchiveFile() {
		log.info("[relocateArchiveFile]");
		for (Flay flay : archiveFlaySource.list()) {
			String yyyyMM = getArchiveFolder(flay);
			File destDir = new File(archivePath, yyyyMM);
			for (Entry<String, List<File>> entry : flay.getFiles().entrySet()) {
				for (File file : entry.getValue()) {
					String parentName = file.getParentFile().getName();
					if (StringUtils.equals(parentName, yyyyMM)) {
						// ok. normal position
					} else {
						log.info("move {} to {}", file, destDir);
						FlayFileHandler.moveFileToDirectory(file, destDir);
					}
				}
			}
		}
	}

	private String getArchiveFolder(Flay flay) {
		return StringUtils.substring(flay.getRelease(), 0, 4) + "-" + StringUtils.substring(flay.getRelease(), 5, 7);
	}

	private void moveWatched() {
		log.info("[moveWatched]");
		for (Flay flay : instanceFlaySource.list()) {
			if (flay.getVideo().getRank() > 0 && flay.getVideo().getPlay() > 0) {
				for (Entry<String, List<File>> entry : flay.getFiles().entrySet()) {
					for (File file : entry.getValue()) {
						if (!file.getPath().startsWith(storagePath)) {
							File destDir = new File(storagePath, flay.getStudio());
							log.info("move {} to {}", file, destDir);
							FlayFileHandler.moveFileToDirectory(file, destDir);
						}
					}
				}
			}
		}
	}

	private void deleteLowerRank() {
		log.info("[deleteLowerRank]");
		for (Flay flay : instanceFlaySource.list()) {
			if (flay.getVideo().getRank() < 0) {
				log.info("lower rank {}", flay.getOpus());
				archiving(flay);
			}
		}
	}

	private void deleteLowerScore() {
		log.info("[deleteLowerScore]");
		long lengthSum = 0;
		long storageSize = storageGbSize * FileUtils.ONE_GB;
		List<Flay> sorted = instanceFlaySource.list().stream()
				.filter(f -> f.getFiles().get(Flay.MOVIE).size() > 0 && f.getVideo().getRank() > 0 && f.getVideo().getPlay() > 0)
				.sorted((f1, f2) -> ScoreCalculator.compare(f2, f1))
				.collect(Collectors.toList());
		for (Flay flay : sorted) {
			lengthSum += flay.getLength();
			if (lengthSum > storageSize) {
				log.info("lower score {} score={} over {}", flay.getOpus(), ScoreCalculator.calc(flay), FlayFileHandler.prettyFileLength(lengthSum));
				archiving(flay);
			}
		}
	}

	private void assembleFlay() {
		log.info("[assembleFlay]");
		for (Flay flay : instanceFlaySource.list()) {
			File delegatePath = getDelegatePath(flay);
			log.debug("assemble {} : {}", flay.getOpus(), delegatePath);
			for (Entry<String, List<File>> entry : flay.getFiles().entrySet()) {
				// String key = entry.getKey();
				List<File> value = entry.getValue();
				for (File file : value) {
					if (!delegatePath.equals(file.getParentFile())) {
						log.info("move [{}] {} to {}", flay.getOpus(), file, delegatePath);
						FlayFileHandler.moveFileToDirectory(file, delegatePath);
					}
				}
			}
		}
	}

	private File getDelegatePath(Flay flay) {
		int movieSize = flay.getFiles().get(Flay.MOVIE).size();
		int coverSize = flay.getFiles().get(Flay.COVER).size();
		
		if (movieSize > 0) {
			String path = flay.getFiles().get(Flay.MOVIE).get(0).getParent();
			if (StringUtils.equalsAny(path, stagePaths)) { // stage 같은 경로에 있으면, 날자 sub폴더
				return new File(path, flay.getRelease().substring(0, 4));
			} else if (StringUtils.equals(path, storagePath)) { // storage 같은 경로에 있으면, studio sub폴더
				return new File(path, flay.getStudio());
			} else if (StringUtils.startsWithAny(path, stagePaths) || StringUtils.startsWith(path, storagePath)) { // 하위에 있으면 현폴더
				return new File(path);
			}
		} else if (coverSize > 0) {
			String path = flay.getFiles().get(Flay.COVER).get(0).getParent();
			if (StringUtils.equalsAny(path, coverPaths)) { // cover 같은 경로에 있으면, 날자 sub폴더 
				return new File(path, flay.getRelease().substring(0, 4));
			} else if (StringUtils.startsWithAny(path, coverPaths)) { // 하위 경로에 있으면, 현폴더
				return new File(path);
			}
		}
		
		log.info("Not determine delegate path, move to Queue. {}", flay.getOpus());
		return new File(queuePath);
	}

	private void deleteEmptyFolder(String...emptyManagedPath) {
		log.info("[deleteEmptyFolder]");
		List<Path> paths = new ArrayList<>();
		for (String dir : emptyManagedPath) {
			try {
				Path start = Paths.get(dir);
				Files.walkFileTree(start, new SimpleFileVisitor<Path>() {

					@Override
					public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
						if (!start.equals(dir) && !Files.newDirectoryStream(dir).iterator().hasNext()) {
							paths.add(dir);
						}
						return super.preVisitDirectory(dir, attrs);
					}});
			} catch (IOException e) {
				throw new IllegalStateException("deleteEmptyFolder walk fail", e);
			}
		}
		for (Path dir : paths) {
			log.info("empty directory delete {}", dir);
			FlayFileHandler.deleteDirectory(dir.toFile());
		}
	}

	private void archiving(Flay flay) {
		String yyyyMM = getArchiveFolder(flay);
		File destDir = new File(archivePath, yyyyMM);
		for (Entry<String, List<File>> entry : flay.getFiles().entrySet()) {
			String key = entry.getKey();
			for (File file : entry.getValue()) {
				if (Flay.COVER.equals(key) || Flay.SUBTI.equals(key)) {
					log.info("move {} to {}", file, destDir);
					FlayFileHandler.moveFileToDirectory(file, destDir);
				} else {
					log.info("delete {}", file);
					FlayFileHandler.deleteFile(file);
				}
			}
		}
		historyService.save(Action.DELETE, flay);
	}
	
	public synchronized void backup() {
		if (StringUtils.isBlank(backupPath)) {
			log.warn("Backup path not set");
			return;
		}
		log.info("[Backup] START {}", backupPath);

		final String CSV_HEADER = "Studio,Opus,Title,Actress,Released,Rank,Fullname";
		final String CSV_FORMAT = "\"%s\",\"%s\",\"%s\",\"%s\",\"%s\",%s,\"%s\"";

		File backupInstanceJarFile = new File(backupPath, BACKUP_INSTANCE_JAR_FILENAME);
		File backupArchiveJarFile  = new File(backupPath, BACKUP_ARCHIVE_JAR_FILENAME);
		File backupRootPath = new File(queuePath, "backup_" + FlayConfig.YYYY_MM_DD_Format.format(new Date()));
		File backupInstanceFilePath = new File(backupRootPath, "instance");
		FlayFileHandler.createDirectory(backupRootPath);
		FlayFileHandler.cleanDirectory(backupRootPath);
		FlayFileHandler.createDirectory(backupInstanceFilePath);

		// video list backup to csv
		Collection<Flay> instanceFlayList = instanceFlaySource.list();
		Collection<Flay>  archiveFlayList =  archiveFlaySource.list();
		List<History>    historyList = historyService.list();

		List<String>    instanceList = new ArrayList<>();
		List<String>     archiveList = new ArrayList<>();

		// instance info
		log.info("[Backup] Write instance csv {} to {}", BACKUP_INSTANCE_CSV_FILENAME, backupRootPath);
		instanceList.add(CSV_HEADER);
		for (Flay flay : instanceFlayList) {
			instanceList.add(String.format(CSV_FORMAT, 
					flay.getStudio(), flay.getOpus(), flay.getTitle(), flay.getActressName(), flay.getRelease(), flay.getVideo().getRank(), flay.getFullname()));
		}
		writeFileWithUTF8BOM(new File(backupRootPath, BACKUP_INSTANCE_CSV_FILENAME), instanceList); 
		
		// archive info
		log.info("[Backup] Write archive  csv {}  to {}", BACKUP_ARCHIVE_CSV_FILENAME, backupRootPath);
		archiveList.add(CSV_HEADER);
		for (Flay flay : archiveFlayList) {
			archiveList.add(String.format(CSV_FORMAT, 
					flay.getStudio(), flay.getOpus(), flay.getTitle(), flay.getActressName(), flay.getRelease(), "", flay.getFullname()));
		}
		for (History history : historyList) {
			String opus = history.getOpus();
			boolean foundInArchive = false;
			for (Flay flay : archiveFlayList) {
				if (flay.getOpus().equals(opus)) {
					foundInArchive = true;
					break;
				}
			}
			if (!foundInArchive)
				archiveList.add(String.format(CSV_FORMAT, "", history.getOpus(), "", "", "", "", history.getDesc()));
		}
		writeFileWithUTF8BOM(new File(backupRootPath, BACKUP_ARCHIVE_CSV_FILENAME),  archiveList);

		// Info folder copy
		log.info("[Backup] Copy Info folder {} to {}", infoPath, backupRootPath);
		FlayFileHandler.copyDirectoryToDirectory(new File(infoPath), backupRootPath);
		
		// Cover, Subtitlea file copy
		log.info("[Backup] Copy Instance file to {}", backupInstanceFilePath);
		for (Flay flay : instanceFlayList) {
			for (File file : flay.getFiles().get(Flay.COVER))
				FlayFileHandler.copyFileToDirectory(file, backupInstanceFilePath);
			for (File file : flay.getFiles().get(Flay.SUBTI))
				FlayFileHandler.copyFileToDirectory(file, backupInstanceFilePath);
		}
//		for (Flay flay : archiveFlayList) {
//			for (File file : flay.getFiles().get(Flay.COVER))
//				FlayFileHandler.copyFileToDirectory(file, backupFilePath);
//			for (File file : flay.getFiles().get(Flay.SUBTI))
//				FlayFileHandler.copyFileToDirectory(file, backupFilePath);
//		}
		
		log.info("[Backup] Compress Instance folder");
		compress(backupRootPath, backupInstanceJarFile);
		
		log.info("[Backup] Compress Archive folder");
		compress(new File(archivePath), backupArchiveJarFile);

		log.info("[Backup] Delete Instance folder {}", backupRootPath);
		FlayFileHandler.deleteDirectory(backupRootPath);
		
		notificationService.announce("Backup", "Flay source");

		log.info("[Backup] END");
	}
	
	private void writeFileWithUTF8BOM(File file, Collection<String> lines) {
		try (BufferedWriter bufferedWriter = Files.newBufferedWriter(file.toPath(), Charset.forName("UTF-8"), StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
			bufferedWriter.write(65279); // UTF-8의 BOM인 "EF BB BF"를 UTF-16BE 로 변환
			for (String line : lines) {
				bufferedWriter.write(line);
				bufferedWriter.newLine();
			}
			bufferedWriter.flush();
		} catch (IOException e) {
			throw new IllegalStateException("Fail to writeFileWithUTF8BOM", e);
		}
	}

	private void compress(File srcPath, File destinationFile) {
		List<String> commands = Arrays.asList("jar", "c0fM", destinationFile.getAbsolutePath(), "-C", srcPath.getAbsolutePath(), ".");
		File logFile = new File(srcPath.getParentFile(), srcPath.getName() + ".log");
		ProcessBuilder builder = new ProcessBuilder(commands);
		builder.redirectOutput(Redirect.to(logFile));
		builder.redirectError(Redirect.INHERIT);
		try {
			log.info("         jar {}", commands);
			Process process = builder.start();
			process.waitFor();
			log.info("         completed {}", FlayFileHandler.prettyFileLength(destinationFile.length()));
		} catch (IOException | InterruptedException e) {
			throw new IllegalStateException("Fail to jar", e);
		}
	}

}

File: server_node/src/services/batch.service.ts
import fs from 'fs';
import path from 'path';
import { config } from '../config';
import { Flay } from '../domain/flay';
import { createHistory } from '../domain/history';
import { getArchiveFlay, getArchiveFlayList, getInstanceFlayList, loadAllFlaySources, reloadInstanceFlaySources } from '../sources/flay-source';
import { historyRepository } from '../sources/history-repository';
import { deleteFile, isEmptyDirectory, moveFileToDirectory } from './flay-file-handler';
import { listOrderByScoreDesc } from './score-calculator';
import { SseMessage, sseSend } from './sse-emitters';

/**
 * 배치 서비스.
 * Java BatchExecutor 대응 - 인스턴스/아카이브 배치, 백업 등
 */

/** 배치 옵션 */
export type BatchOption = 'S' | 'R';

/** 배치 오퍼레이션 */
export type BatchOperation = 'I' | 'A' | 'B';

/** 런타임 옵션 상태 */
let deleteLowerScore = config.flay.deleteLowerScore;

/**
 * SSE를 통해 배치 로그를 전송한다.
 */
function batchLogger(message: string): void {
  console.log(`[Batch] ${message}`);
  const msg: SseMessage = { type: 'Batch', message };
  sseSend(msg);
}

/**
 * SSE를 통해 공지 알림을 전송한다.
 */
function noticeLogger(message: string): void {
  console.log(`[Batch] ${message}`);
  const msg: SseMessage = { type: 'Notice', message };
  sseSend(msg);
}

/** 배치 옵션 조회 */
export function getOption(option: BatchOption): boolean {
  switch (option) {
    case 'S':
      return deleteLowerScore;
    default:
      throw new Error(`알 수 없는 배치 옵션: ${option}`);
  }
}

/** 배치 옵션 토글 */
export function toggleOption(option: BatchOption): boolean {
  switch (option) {
    case 'S':
      deleteLowerScore = !deleteLowerScore;
      return deleteLowerScore;
    default:
      throw new Error(`알 수 없는 배치 옵션: ${option}`);
  }
}

/** 배치 실행 (비동기) */
export function startBatch(operation: BatchOperation): void {
  // 비동기로 실행 (Node.js에서는 setImmediate로 메인 스레드 해제)
  setImmediate(() => {
    try {
      switch (operation) {
        case 'I':
          instanceBatch();
          break;
        case 'A':
          archiveBatch();
          break;
        case 'B':
          backup();
          break;
        default:
          throw new Error(`알 수 없는 배치 오퍼레이션: ${operation}`);
      }
    } catch (err: any) {
      console.error(`[Batch] 오류: ${err.message}`);
      batchLogger(`오류: ${err.message}`);
    }
  });
}

/** 배치 사전 검사 */
export function checkBatch(operation: BatchOperation): Record<string, Flay[]> {
  switch (operation) {
    case 'I':
      return instanceCheck();
    default:
      throw new Error(`알 수 없는 배치 오퍼레이션: ${operation}`);
  }
}

/** 소스 리로드 */
export function reload(): void {
  batchLogger('[reload] Start');
  reloadInstanceFlaySources();
  batchLogger('[reload] End');
  noticeLogger('reload Completed');
}

// ─── 내부 구현 ───

/** rank < 0인 Flay 목록 */
function listLowerRank(): Flay[] {
  return getInstanceFlayList().filter((flay) => flay.video.rank < 0);
}

/** 용량 초과 시 low score Flay 목록 */
function listLowerScore(): Flay[] {
  const storageSize = config.flay.storageLimit * 1024 * 1024 * 1024;
  const sorted = listOrderByScoreDesc(getInstanceFlayList());
  const result: Flay[] = [];
  let lengthSum = 0;

  for (const flay of sorted) {
    lengthSum += flay.length;
    if (lengthSum > storageSize) {
      result.push(flay);
    }
  }

  return result;
}

/** 인스턴스 배치 사전 검사 */
function instanceCheck(): Record<string, Flay[]> {
  return {
    rank: listLowerRank(),
    score: listLowerScore(),
  };
}

/** 인스턴스 배치 실행 */
function instanceBatch(): void {
  batchLogger('[instanceBatch] Start');

  // rank < 0 삭제
  batchLogger('[deleteLowerRank]');
  for (const flay of listLowerRank()) {
    archiving(flay, 'delete Lower Rank Video');
  }

  // low score 삭제
  if (deleteLowerScore) {
    batchLogger('[deleteLowerScore]');
    for (const flay of listLowerScore()) {
      archiving(flay, 'delete Lower Score Video');
    }
  }

  // 파일 재배치
  assembleFlay();

  // 빈 폴더 삭제
  const pathsToClean = [...config.flay.stagePaths, config.flay.coverPath, config.flay.storagePath];
  deleteEmptyFolders(pathsToClean);

  // 소스 리로드
  reloadInstanceFlaySources();

  batchLogger('[instanceBatch] End');
  noticeLogger('instanceBatch Completed');
}

/** 아카이브 배치 실행 */
function archiveBatch(): void {
  batchLogger('[archiveBatch] Start');
  batchLogger('[relocateArchiveFile]');

  for (const flay of getArchiveFlayList()) {
    const yyyyMM = getArchiveFolderName(flay);
    const destDir = path.join(config.flay.archivePath, yyyyMM);

    for (const [, fileList] of Object.entries(flay.files)) {
      for (const filePath of fileList) {
        const parentName = path.basename(path.dirname(filePath));
        if (parentName !== yyyyMM) {
          batchLogger(`move ${filePath} to ${destDir}`);
          moveFileToDirectory(filePath, destDir);
        }
      }
    }
  }

  deleteEmptyFolders([config.flay.archivePath]);
  loadAllFlaySources();

  batchLogger('[archiveBatch] End');
  noticeLogger('archiveBatch Completed');
}

/**
 * Flay 파일을 적절한 위치로 재배치한다.
 * Java BatchExecutor.assembleFlay() 대응
 */
function assembleFlay(): void {
  batchLogger('[assembleFlay]');

  for (const flay of getInstanceFlayList()) {
    if (flay.archive) continue;

    // 커버/자막이 없으면 아카이브에서 찾기
    recoverFromArchive(flay);

    const delegatePath = getDelegatePath(flay);

    for (const [, fileList] of Object.entries(flay.files)) {
      for (const filePath of fileList) {
        const parentDir = path.dirname(filePath);
        if (path.resolve(parentDir) !== path.resolve(delegatePath)) {
          batchLogger(`move ${path.basename(filePath)} => ${delegatePath}`);
          moveFileToDirectory(filePath, delegatePath);
        }
      }
    }
  }
}

/**
 * 아카이브에서 커버/자막 파일을 복원한다.
 */
function recoverFromArchive(flay: Flay): void {
  // 커버 파일이 없으면 아카이브에서 찾기
  if (!flay.files.cover || flay.files.cover.length === 0) {
    try {
      const archiveFlay = getArchiveFlay(flay.opus);
      if (archiveFlay.files.cover && archiveFlay.files.cover.length > 0) {
        flay.files.cover.push(archiveFlay.files.cover[0]);
        batchLogger(`add Cover ${archiveFlay.files.cover[0]}`);
      }
    } catch {
      // 아카이브에도 없으면 무시
    }
  }

  // 자막 파일이 없으면 아카이브에서 찾기
  if (!flay.files.subtitles || flay.files.subtitles.length === 0) {
    try {
      const archiveFlay = getArchiveFlay(flay.opus);
      if (archiveFlay.files.subtitles && archiveFlay.files.subtitles.length > 0) {
        flay.files.subtitles.push(...archiveFlay.files.subtitles);
        batchLogger(`add subtitles ${archiveFlay.files.subtitles.join(', ')}`);
      }
    } catch {
      // 무시
    }
  }
}

/**
 * Flay가 위치해야 할 폴더를 결정한다.
 * Java BatchExecutor.getDelegatePath() 대응
 */
function getDelegatePath(flay: Flay): string {
  const movieCount = flay.files.movie?.length || 0;
  const coverCount = flay.files.cover?.length || 0;

  if (movieCount > 0) {
    const movieDir = path.dirname(flay.files.movie[0]);
    const rank = flay.video.rank;

    if (rank > 0) {
      // storage에 [studio]로
      return path.join(config.flay.storagePath, flay.studio);
    } else if (rank === 0) {
      // 같은 디스크의 stage에 [연도]로
      const root = path.parse(movieDir).root;
      const stagePath = config.flay.stagePaths.find((p) => p.startsWith(root.substring(0, 1)));
      if (stagePath) {
        return path.join(stagePath, flay.release.substring(0, 4));
      }
    }
  } else if (coverCount > 0) {
    // 비디오 없는 파일은 cover로
    return path.join(config.flay.coverPath, flay.release.substring(0, 4));
  }

  batchLogger(`Not determine delegate path, return to Queue path. ${flay.opus}`);
  return config.flay.queuePath;
}

/**
 * 아카이빙: cover/subtitles는 아카이브로 이동, 나머지는 삭제
 * Java BatchExecutor.archiving() 대응
 */
function archiving(flay: Flay, reason: string): void {
  const yyyyMM = getArchiveFolderName(flay);
  const archiveDir = path.join(config.flay.archivePath, yyyyMM);

  for (const [key, fileList] of Object.entries(flay.files)) {
    for (const filePath of fileList) {
      if (key === 'cover' || key === 'subtitles') {
        batchLogger(`will be move   ${filePath} to ${archiveDir}`);
        moveFileToDirectory(filePath, archiveDir);
      } else if (key !== 'candidate') {
        batchLogger(`will be delete ${filePath}`);
        deleteFile(filePath);
      }
    }
  }

  flay.archive = true;
  historyRepository.save(createHistory(flay.opus, 'DELETE', reason));
}

/** 아카이브 폴더명 (yyyy-MM 형식) */
function getArchiveFolderName(flay: Flay): string {
  return flay.release.substring(0, 4) + '-' + flay.release.substring(5, 7);
}

/** 빈 폴더 삭제 */
function deleteEmptyFolders(dirs: string[]): void {
  batchLogger('[deleteEmptyFolder]');
  for (const dir of dirs) {
    if (!fs.existsSync(dir)) continue;
    batchLogger(`  scanning...   ${dir}`);
    cleanEmptySubdirs(dir, dir);
  }
}

/** 재귀적으로 빈 하위 디렉토리를 삭제한다 */
function cleanEmptySubdirs(rootDir: string, currentDir: string): void {
  if (!fs.existsSync(currentDir)) return;
  const entries = fs.readdirSync(currentDir, { withFileTypes: true });

  for (const entry of entries) {
    if (entry.isDirectory()) {
      const subDir = path.join(currentDir, entry.name);
      cleanEmptySubdirs(rootDir, subDir);

      if (isEmptyDirectory(subDir)) {
        batchLogger(`    empty directory delete ${subDir}`);
        fs.rmdirSync(subDir);
      }
    }
  }
}

/**
 * 백업 실행.
 * Java BatchExecutor.backup() 대응
 * instance/archive CSV와 Info 폴더를 백업한다.
 */
function backup(): void {
  const backupPath = config.flay.backupPath;
  if (!fs.existsSync(backupPath)) {
    console.warn('[Batch] 백업 경로가 존재하지 않습니다.');
    return;
  }

  batchLogger(`[Backup] START ${backupPath}`);

  const CSV_HEADER = 'Studio,Opus,Title,Actress,Released,Rank,Fullname';
  const instanceFlayList = getInstanceFlayList();
  const archiveFlayList = getArchiveFlayList();

  // Instance CSV
  const instanceCsvFilename = config.flay.backup.instanceCsvFilename;
  const instanceCsvPath = path.join(backupPath, instanceCsvFilename);
  const instanceLines = [CSV_HEADER];
  for (const flay of instanceFlayList) {
    const actress = flay.actressList.join(',');
    const fullname = `[${flay.studio}][${flay.opus}][${flay.title}][${actress}][${flay.release}]`;
    instanceLines.push(`"${flay.studio}","${flay.opus}","${flay.title}","${actress}","${flay.release}",${flay.video.rank},"${fullname}"`);
  }
  writeFileWithUtf8Bom(instanceCsvPath, instanceLines);
  batchLogger(`[Backup] Write instance csv ${instanceCsvFilename}`);

  // Archive CSV
  const archiveCsvFilename = config.flay.backup.archiveCsvFilename;
  const archiveCsvPath = path.join(backupPath, archiveCsvFilename);
  const archiveLines = [CSV_HEADER];
  for (const flay of archiveFlayList) {
    const actress = flay.actressList.join(',');
    const fullname = `[${flay.studio}][${flay.opus}][${flay.title}][${actress}][${flay.release}]`;
    archiveLines.push(`"${flay.studio}","${flay.opus}","${flay.title}","${actress}","${flay.release}",,"${fullname}"`);
  }
  writeFileWithUtf8Bom(archiveCsvPath, archiveLines);
  batchLogger(`[Backup] Write archive csv ${archiveCsvFilename}`);

  // Info 폴더 복사
  const infoSrc = config.flay.infoPath;
  const infoDest = path.join(backupPath, 'Info');
  if (fs.existsSync(infoSrc)) {
    copyDirectoryRecursive(infoSrc, infoDest);
    batchLogger(`[Backup] Copy Info folder ${infoSrc} to ${infoDest}`);
  }

  batchLogger('[Backup] END');
  noticeLogger('backup Completed');
}

/** UTF-8 BOM 포함 파일 쓰기 */
function writeFileWithUtf8Bom(filePath: string, lines: string[]): void {
  const content = '\uFEFF' + lines.join('\n') + '\n';
  fs.writeFileSync(filePath, content, 'utf-8');
}

/** 디렉토리를 재귀적으로 복사한다 */
function copyDirectoryRecursive(src: string, dest: string): void {
  if (!fs.existsSync(dest)) fs.mkdirSync(dest, { recursive: true });

  const entries = fs.readdirSync(src, { withFileTypes: true });
  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);
    if (entry.isDirectory()) {
      copyDirectoryRecursive(srcPath, destPath);
    } else {
      fs.copyFileSync(srcPath, destPath);
    }
  }
}
